@hmlnarik Thank you for the PR.  When I played with this I noticed there's a `ConcurrentModificationException` when running with `jpa+infinispan`
`WARN  [org.infinispan.CLUSTER] (jgroups-11,node-2) ISPN000071: Caught exception when handling command org.infinispan.manager.impl.ReplicableManagerFunctionCommand@3bc68b73
liquibase.exception.UnexpectedLiquibaseException: java.util.ConcurrentModificationException
	at liquibase.datatype.DataTypeFactory.register(DataTypeFactory.java:81)
	at org.keycloak.connections.jpa.updater.liquibase.conn.DefaultLiquibaseConnectionProvider.baseLiquibaseInitialization(DefaultLiquibaseConnectionProvider.java:112)
	at org.keycloak.connections.jpa.updater.liquibase.conn.DefaultLiquibaseConnectionProvider.create(DefaultLiquibaseConnectionProvider.java:70)
	at org.keycloak.connections.jpa.updater.liquibase.conn.DefaultLiquibaseConnectionProvider.create(DefaultLiquibaseConnectionProvider.java:55)
	at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:315)
	at org.keycloak.connections.jpa.updater.liquibase.LiquibaseJpaUpdaterProvider.getLiquibaseForKeycloakUpdate(LiquibaseJpaUpdaterProvider.java:289)
	at org.keycloak.connections.jpa.updater.liquibase.LiquibaseJpaUpdaterProvider.validate(LiquibaseJpaUpdaterProvider.java:224)
	at org.keycloak.connections.jpa.DefaultJpaConnectionProviderFactory.migration(DefaultJpaConnectionProviderFactory.java:301)
	at org.keycloak.connections.jpa.DefaultJpaConnectionProviderFactory.lambda$lazyInit$0(DefaultJpaConnectionProviderFactory.java:182)
	at org.keycloak.models.utils.KeycloakModelUtils.suspendJtaTransaction(KeycloakModelUtils.java:771)
	at org.keycloak.connections.jpa.DefaultJpaConnectionProviderFactory.lazyInit(DefaultJpaConnectionProviderFactory.java:133)
	at org.keycloak.connections.jpa.DefaultJpaConnectionProviderFactory.create(DefaultJpaConnectionProviderFactory.java:81)
	at org.keycloak.connections.jpa.DefaultJpaConnectionProviderFactory.create(DefaultJpaConnectionProviderFactory.java:59)
	at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:315)
	at org.keycloak.models.jpa.session.JpaUserSessionPersisterProviderFactory.create(JpaUserSessionPersisterProviderFactory.java:38)
	at org.keycloak.models.jpa.session.JpaUserSessionPersisterProviderFactory.create(JpaUserSessionPersisterProviderFactory.java:32)
	at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:315)
	at org.keycloak.models.sessions.infinispan.initializer.OfflinePersistentUserSessionLoader.loadSessions(OfflinePersistentUserSessionLoader.java:102)
	at org.keycloak.models.sessions.infinispan.initializer.OfflinePersistentUserSessionLoader.loadSessions(OfflinePersistentUserSessionLoader.java:37)
	at org.keycloak.models.sessions.infinispan.initializer.SessionInitializerWorker$1.run(SessionInitializerWorker.java:71)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:250)
	at org.keycloak.models.sessions.infinispan.initializer.SessionInitializerWorker.apply(SessionInitializerWorker.java:67)
	at org.keycloak.models.sessions.infinispan.initializer.SessionInitializerWorker.apply(SessionInitializerWorker.java:34)
	at org.infinispan.manager.impl.ReplicableManagerFunctionCommand.invokeAsync(ReplicableManagerFunctionCommand.java:39)
	at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler$ReplicableCommandRunner.run(GlobalInboundInvocationHandler.java:258)
	at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleReplicableCommand(GlobalInboundInvocationHandler.java:177)
	at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleFromCluster(GlobalInboundInvocationHandler.java:115)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processRequest(JGroupsTransport.java:1378)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1302)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.access$300(JGroupsTransport.java:131)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1445)
	at org.jgroups.JChannel.up(JChannel.java:775)
	at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:900)
	at org.jgroups.protocols.FRAG3.up(FRAG3.java:165)
	at org.jgroups.protocols.FlowControl.up(FlowControl.java:351)
	at org.jgroups.protocols.FlowControl.up(FlowControl.java:359)
	at org.jgroups.protocols.pbcast.GMS.up(GMS.java:868)
	at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:243)
	at org.jgroups.protocols.UNICAST3.deliverMessage(UNICAST3.java:1049)
	at org.jgroups.protocols.UNICAST3.addMessage(UNICAST3.java:772)
	at org.jgroups.protocols.UNICAST3.handleDataReceived(UNICAST3.java:753)
	at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:405)
	at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:592)
	at org.jgroups.protocols.VERIFY_SUSPECT.up(VERIFY_SUSPECT.java:132)
	at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:186)
	at org.jgroups.protocols.FD_SOCK.up(FD_SOCK.java:254)
	at org.jgroups.protocols.MERGE3.up(MERGE3.java:281)
	at org.jgroups.protocols.Discovery.up(Discovery.java:300)
	at org.jgroups.protocols.TP.passMessageUp(TP.java:1382)
	at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:87)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.ConcurrentModificationException
	at java.base/java.util.ArrayList.sort(ArrayList.java:1752)
	at java.base/java.util.Collections.sort(Collections.java:179)
	at liquibase.datatype.DataTypeFactory.register(DataTypeFactory.java:78)
	... 52 more
`

--- a/model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java
+++ b/model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java
@@ -312,6 +312,7 @@ public class DefaultInfinispanConnectionProviderFactory implements InfinispanCon
 
         Configuration replicationEvictionCacheConfiguration = replicationConfigBuilder.build();
         cacheManager.defineConfiguration(InfinispanConnectionProvider.WORK_CACHE_NAME, replicationEvictionCacheConfiguration);
+        cacheManager.getCache(InfinispanConnectionProvider.WORK_CACHE_NAME, true);
 
         long realmRevisionsMaxEntries = cacheManager.getCache(InfinispanConnectionProvider.REALM_CACHE_NAME).getCacheConfiguration().memory().size();
         realmRevisionsMaxEntries = realmRevisionsMaxEntries > 0

--- a/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/initializer/InfinispanCacheInitializer.java
+++ b/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/initializer/InfinispanCacheInitializer.java
@@ -158,7 +158,7 @@ public class InfinispanCacheInitializer extends BaseCacheInitializer {
 
                     completableFuture = clusterExecutor.submitConsumer(worker, (address, workerResult, throwable) -> {
                         log.tracef("Calling triConsumer on address %s, throwable message: %s, segment: %s", address, throwable == null ? "null" : throwable.getMessage(),
-                                workerResult.getSegment());
+                                workerResult == null ? null : workerResult.getSegment());
 
                         if (throwable != null) {
                             throw new CacheException(throwable);

--- a/model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/conn/DefaultLiquibaseConnectionProvider.java
+++ b/model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/conn/DefaultLiquibaseConnectionProvider.java
@@ -48,6 +48,7 @@ import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.KeycloakSessionFactory;
 
 import java.sql.Connection;
+import java.util.concurrent.atomic.AtomicBoolean;
 import liquibase.changelog.ChangeLogHistoryServiceFactory;
 
 /**
@@ -61,15 +62,18 @@ public class DefaultLiquibaseConnectionProvider implements LiquibaseConnectionPr
 
     private int indexCreationThreshold;
 
-    private volatile boolean initialized = false;
+    private static final AtomicBoolean INITIALIZATION = new AtomicBoolean(false);
     
     @Override
     public LiquibaseConnectionProvider create(KeycloakSession session) {
-        if (!initialized) {
-            synchronized (this) {
-                if (!initialized) {
+        if (! INITIALIZATION.get()) {
+            // We need critical section synchronized on some static final field, otherwise
+            // e.g. several Undertows or parallel model tests could attempt initializing Liquibase
+            // in the same JVM at the same time which leads to concurrency failures
+            synchronized (INITIALIZATION) {
+                if (! INITIALIZATION.get()) {
                     baseLiquibaseInitialization();
-                    initialized = true;
+                    INITIALIZATION.set(true);
                 }
             }
         }

--- a/model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProvider.java
+++ b/model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProvider.java
@@ -42,6 +42,7 @@ import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
 import java.util.function.Function;
 import java.util.stream.Collectors;
@@ -253,6 +254,7 @@ public class JpaUserSessionPersisterProvider implements UserSessionPersisterProv
 
         List<PersistentUserSessionAdapter> result = closing(paginateQuery(query, firstResult, maxResults).getResultStream()
                 .map(this::toAdapter))
+                .filter(Objects::nonNull)
                 .collect(Collectors.toList());
 
         Map<String, PersistentUserSessionAdapter> sessionsById = result.stream()
@@ -290,6 +292,9 @@ public class JpaUserSessionPersisterProvider implements UserSessionPersisterProv
 
     private PersistentUserSessionAdapter toAdapter(PersistentUserSessionEntity entity) {
         RealmModel realm = session.realms().getRealm(entity.getRealmId());
+        if (realm == null) {    // Realm has been deleted concurrently, ignore the entity
+            return null;
+        }
         return toAdapter(realm, entity);
     }
 

--- a/model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java
+++ b/model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java
@@ -38,7 +38,7 @@ import java.io.Serializable;
         @NamedQuery(name="updateUserSessionLastSessionRefresh", query="update PersistentUserSessionEntity sess set lastSessionRefresh = :lastSessionRefresh where sess.realmId = :realmId" +
                 " AND sess.offline = :offline AND sess.userSessionId IN (:userSessionIds)"),
         @NamedQuery(name="findUserSessionsCount", query="select count(sess) from PersistentUserSessionEntity sess where sess.offline = :offline"),
-        @NamedQuery(name="findUserSessions", query="select sess from PersistentUserSessionEntity sess where sess.offline = :offline" +
+        @NamedQuery(name="findUserSessions", query="select sess from PersistentUserSessionEntity sess, RealmEntity realm where realm.id = sess.realmId AND sess.offline = :offline" +
                 " AND (sess.createdOn > :lastCreatedOn OR (sess.createdOn = :lastCreatedOn AND sess.userSessionId > :lastSessionId))" +
                 " order by sess.createdOn,sess.userSessionId")
 

--- a/testsuite/model/src/main/java/org/keycloak/testsuite/model/Config.java
+++ b/testsuite/model/src/main/java/org/keycloak/testsuite/model/Config.java
@@ -34,7 +34,16 @@ public class Config implements ConfigProvider {
 
     private final Properties systemProperties = new SystemEnvProperties();
 
-    private final Map<String, String> properties = new HashMap<>();
+    private final ThreadLocal<Map<String, String>> properties = new ThreadLocal<Map<String, String>>() {
+        @Override
+        protected Map<String, String> initialValue() {
+            return new HashMap<>();
+        }
+    };
+
+    void reset() {
+        properties.remove();
+    }
 
     public class SpiConfig {
 
@@ -54,9 +63,9 @@ public class Config implements ConfigProvider {
 
         public SpiConfig config(String key, String value) {
             if (value == null) {
-                properties.remove(prefix + key);
+                properties.get().remove(prefix + key);
             } else {
-                properties.put(prefix + key, value);
+                properties.get().put(prefix + key, value);
             }
             return this;
         }
@@ -78,9 +87,9 @@ public class Config implements ConfigProvider {
 
         public ProviderConfig config(String key, String value) {
             if (value == null) {
-                properties.remove(prefix + key);
+                properties.get().remove(prefix + key);
             } else {
-                properties.put(prefix + key, value);
+                properties.get().put(prefix + key, value);
             }
             return this;
         }
@@ -103,7 +112,7 @@ public class Config implements ConfigProvider {
 
         @Override
         public String get(String key, String defaultValue) {
-            String v = replaceProperties(properties.get(prefix + key));
+            String v = replaceProperties(properties.get().get(prefix + key));
             if (v == null || v.isEmpty()) {
                 v = System.getProperty("keycloak." + prefix + key, defaultValue);
             }
@@ -124,7 +133,7 @@ public class Config implements ConfigProvider {
 
     @Override
     public String getProvider(String spiName) {
-        return properties.get(spiName + ".provider");
+        return properties.get().get(spiName + ".provider");
     }
 
     private String replaceProperties(String value) {
@@ -147,7 +156,7 @@ public class Config implements ConfigProvider {
 
     @Override
     public String toString() {
-        return properties.entrySet().stream()
+        return properties.get().entrySet().stream()
           .sorted((e1, e2) -> e1.getKey().compareTo(e2.getKey()))
           .map(e -> e.getKey() + " = " + e.getValue())
           .collect(Collectors.joining("\n    "));

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java
@@ -33,6 +33,8 @@ import org.keycloak.models.KeycloakSessionFactory;
 import org.keycloak.models.RealmModel;
 import org.keycloak.models.RealmSpi;
 import org.keycloak.models.RoleSpi;
+import org.keycloak.models.ServerInfoProviderFactory;
+import org.keycloak.models.ServerInfoSpi;
 import org.keycloak.models.UserLoginFailureSpi;
 import org.keycloak.models.UserSessionSpi;
 import org.keycloak.models.UserSpi;
@@ -42,21 +44,28 @@ import org.keycloak.provider.Provider;
 import org.keycloak.provider.ProviderFactory;
 import org.keycloak.provider.ProviderManager;
 import org.keycloak.provider.Spi;
-import org.keycloak.services.DefaultKeycloakSession;
 import org.keycloak.services.DefaultKeycloakSessionFactory;
+import org.keycloak.timer.TimerSpi;
 import com.google.common.collect.ImmutableSet;
+import java.util.Arrays;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 import java.util.stream.Stream;
 import org.hamcrest.Matchers;
 import org.jboss.logging.Logger;
@@ -70,13 +79,10 @@ import org.junit.Rule;
 import org.junit.rules.TestRule;
 import org.junit.runner.Description;
 import org.junit.runners.model.Statement;
-import org.keycloak.timer.TimerSpi;
-import org.keycloak.models.ServerInfoProviderFactory;
-import org.keycloak.models.ServerInfoSpi;
 
 /**
  * Base of testcases that operate on session level. The tests derived from this class
- * will have access to a shared {@link KeycloakSessionFactory} in the {@link #FACTORY}
+ * will have access to a shared {@link KeycloakSessionFactory} in the {@link #LOCAL_FACTORY}
  * field that can be used to obtain a session and e.g. start / stop transaction.
  * <p>
  * This class expects {@code keycloak.model.parameters} system property to contain
@@ -90,7 +96,9 @@ import org.keycloak.models.ServerInfoSpi;
 public abstract class KeycloakModelTest {
 
     private static final Logger LOG = Logger.getLogger(KeycloakModelParameters.class);
+    private static final AtomicInteger FACTORY_COUNT = new AtomicInteger();
     protected final Logger log = Logger.getLogger(getClass());
+    private static final List<String> MAIN_THREAD_NAMES = Arrays.asList("main", "Time-limited test");
 
     @ClassRule
     public static final TestRule GUARANTEE_REQUIRED_FACTORY = new TestRule() {
@@ -104,8 +112,8 @@ public abstract class KeycloakModelTest {
             }
             List<Class<? extends Provider>> notFound = st
               .filter(rp -> rp.only().length == 0 
-                ? FACTORY.getProviderFactory(rp.value()) == null
-                : Stream.of(rp.only()).allMatch(provider -> FACTORY.getProviderFactory(rp.value(), provider) == null))
+                ? getFactory().getProviderFactory(rp.value()) == null
+                : Stream.of(rp.only()).allMatch(provider -> getFactory().getProviderFactory(rp.value(), provider) == null))
               .map(RequireProvider::value)
               .collect(Collectors.toList());
             Assume.assumeThat("Some required providers not found", notFound, Matchers.empty());
@@ -138,7 +146,7 @@ public abstract class KeycloakModelTest {
                 String[] only = rpInner.only();
 
                 if (only.length == 0) {
-                    if (FACTORY.getProviderFactory(providerClass) == null) {
+                    if (getFactory().getProviderFactory(providerClass) == null) {
                         return new Statement() {
                             @Override
                             public void evaluate() throws Throwable {
@@ -147,7 +155,7 @@ public abstract class KeycloakModelTest {
                         };
                     }
                 } else {
-                    boolean notFoundAny = Stream.of(only).allMatch(provider -> FACTORY.getProviderFactory(providerClass, provider) == null);
+                    boolean notFoundAny = Stream.of(only).allMatch(provider -> getFactory().getProviderFactory(providerClass, provider) == null);
                     if (notFoundAny) {
                         return new Statement() {
                             @Override
@@ -192,10 +200,14 @@ public abstract class KeycloakModelTest {
       .build();
 
     protected static final List<KeycloakModelParameters> MODEL_PARAMETERS;
-    protected static Config CONFIG;
-    protected static DefaultKeycloakSessionFactory FACTORY;
+    protected static final Config CONFIG = new Config();
+    private static volatile KeycloakSessionFactory DEFAULT_FACTORY;
+    private static final ThreadLocal<KeycloakSessionFactory> LOCAL_FACTORY = new ThreadLocal<>();
+    protected static boolean USE_DEFAULT_FACTORY = false;
 
     static {
+        org.keycloak.Config.init(CONFIG);
+
         KeycloakModelParameters basicParameters = new KeycloakModelParameters(ALLOWED_SPIS, ALLOWED_FACTORIES);
         MODEL_PARAMETERS = Stream.concat(
           Stream.of(basicParameters),
@@ -210,13 +222,23 @@ public abstract class KeycloakModelTest {
           .collect(Collectors.toList());
 
         reinitializeKeycloakSessionFactory();
+        DEFAULT_FACTORY = getFactory();
     }
 
-    public static DefaultKeycloakSessionFactory createKeycloakSessionFactory() {
-        CONFIG = new Config();
+    /**
+     * Creates a fresh initialized {@link KeycloakSessionFactory}. The returned factory uses configuration
+     * local to the thread that calls this method, allowing for per-thread customization. This in turn allows
+     * testing of several parallel session factories which can be used to simulate several servers
+     * running in parallel.
+     * @return
+     */
+    public static KeycloakSessionFactory createKeycloakSessionFactory() {
+        int factoryIndex = FACTORY_COUNT.incrementAndGet();
+        String threadName = Thread.currentThread().getName();
+        CONFIG.reset();
         MODEL_PARAMETERS.forEach(m -> m.updateConfig(CONFIG));
-        LOG.debug("Using the following configuration:\n    " + CONFIG);
-        org.keycloak.Config.init(CONFIG);
+
+        LOG.debugf("Creating factory %d in %s using the following configuration:\n    %s", factoryIndex, threadName, CONFIG);
 
         DefaultKeycloakSessionFactory res = new DefaultKeycloakSessionFactory() {
             @Override
@@ -237,17 +259,96 @@ public abstract class KeycloakModelTest {
             private boolean isFactoryAllowed(ProviderFactory factory) {
                 return MODEL_PARAMETERS.stream().anyMatch(p -> p.isFactoryAllowed(factory));
             }
+
+            @Override
+            public String toString() {
+                return "KeycloakSessionFactory " + factoryIndex + " (from " + threadName + " thread)";
+            }
         };
         res.init();
         res.publish(new PostMigrationEvent());
         return res;
     }
 
-    public static void reinitializeKeycloakSessionFactory() {
-        if (FACTORY != null) {
-            FACTORY.close();
+    /**
+     * Closes and initializes new {@link #LOCAL_FACTORY}. This has the same effect as server restart in full-blown server scenario.
+     */
+    public static synchronized void reinitializeKeycloakSessionFactory() {
+        closeKeycloakSessionFactory();
+        setFactory(createKeycloakSessionFactory());
+    }
+
+    public static synchronized void closeKeycloakSessionFactory() {
+        KeycloakSessionFactory f = getFactory();
+        setFactory(null);
+        if (f != null) {
+            LOG.debugf("Closing %s", f);
+            f.close();
         }
-        FACTORY = createKeycloakSessionFactory();
+    }
+
+    /**
+     * Runs the given {@code task} in {@code numThreads} parallel threads, each thread operating
+     * in the context of a fresh {@link KeycloakSessionFactory} independent of each other thread.
+     *
+     * @see #inIndependentFactory
+     *
+     * @param numThreads
+     * @param timeoutSeconds
+     * @param task
+     * @throws InterruptedException
+     */
+    public static void inIndependentFactories(int numThreads, int timeoutSeconds, Runnable task) throws InterruptedException {
+        ExecutorService es = Executors.newFixedThreadPool(numThreads);
+        try {
+            Callable<?> independentTask = () -> inIndependentFactory(() -> { task.run(); return null; });
+            es.invokeAll(
+              IntStream.range(0, numThreads)
+                .mapToObj(i -> independentTask)
+                .collect(Collectors.toList()),
+              timeoutSeconds, TimeUnit.SECONDS
+            );
+        } finally {
+            es.shutdownNow();
+        }
+    }
+
+    /**
+     * Runs the given {@code task} in a context of a fresh {@link KeycloakSessionFactory} which is created before
+     * running the task and destroyed afterwards.
+     * @return
+     */
+    public static <T> T inIndependentFactory(Callable<T> task) {
+        if (USE_DEFAULT_FACTORY) {
+            throw new IllegalStateException("USE_DEFAULT_FACTORY must be false to use an independent factory");
+        }
+        KeycloakSessionFactory original = getFactory();
+        KeycloakSessionFactory factory = createKeycloakSessionFactory();
+        try {
+            setFactory(factory);
+            return task.call();
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
+        } finally {
+            closeKeycloakSessionFactory();
+            setFactory(original);
+        }
+    }
+
+    private static boolean isOnMainThread() {
+        return MAIN_THREAD_NAMES.contains(Thread.currentThread().getName());
+    }
+
+    protected static KeycloakSessionFactory getFactory() {
+        return (USE_DEFAULT_FACTORY || isOnMainThread()) ? DEFAULT_FACTORY : LOCAL_FACTORY.get();
+    }
+
+    private static void setFactory(KeycloakSessionFactory factory) {
+        if (USE_DEFAULT_FACTORY || isOnMainThread()) {
+            DEFAULT_FACTORY = factory;
+        } else {
+            LOCAL_FACTORY.set(factory);
+        }
     }
 
     @BeforeClass
@@ -263,12 +364,16 @@ public abstract class KeycloakModelTest {
 
     @Before
     public void createEnvironment() {
-        KeycloakModelUtils.runJobInTransaction(FACTORY, this::createEnvironment);
+        USE_DEFAULT_FACTORY = isUseSameKeycloakSessionFactoryForAllThreads();
+        KeycloakModelUtils.runJobInTransaction(getFactory(), this::createEnvironment);
     }
 
     @After
     public void cleanEnvironment() {
-        KeycloakModelUtils.runJobInTransaction(FACTORY, this::cleanEnvironment);
+        if (getFactory() == null) {
+            reinitializeKeycloakSessionFactory();
+        }
+        KeycloakModelUtils.runJobInTransaction(getFactory(), this::cleanEnvironment);
     }
 
     protected <T> Stream<T> getParameters(Class<T> clazz) {
@@ -280,7 +385,7 @@ public abstract class KeycloakModelTest {
     }
 
     protected <T> void inRolledBackTransaction(T parameter, BiConsumer<KeycloakSession, T> what) {
-        KeycloakSession session = new DefaultKeycloakSession(FACTORY);
+        KeycloakSession session = getFactory().create();
         session.getTransactionManager().begin();
 
         what.accept(session, parameter);
@@ -297,12 +402,12 @@ public abstract class KeycloakModelTest {
     }
 
     protected <R> R inComittedTransaction(Function<KeycloakSession, R> what) {
-        return inComittedTransaction(1, (a,b) -> what.apply(a));
+        return inComittedTransaction(1, (a,b) -> what.apply(a), null, null);
     }
 
     protected <T, R> R inComittedTransaction(T parameter, BiFunction<KeycloakSession, T, R> what, BiConsumer<KeycloakSession, T> onCommit, BiConsumer<KeycloakSession, T> onRollback) {
         AtomicReference<R> res = new AtomicReference<>();
-        KeycloakModelUtils.runJobInTransaction(FACTORY, session -> {
+        KeycloakModelUtils.runJobInTransaction(getFactory(), session -> {
             session.getTransactionManager().enlistAfterCompletion(new AbstractKeycloakTransaction() {
                 @Override
                 protected void commitImpl() {
@@ -319,6 +424,15 @@ public abstract class KeycloakModelTest {
         return res.get();
     }
 
+    /**
+     * Convenience method for {@link #inComittedTransaction(java.util.function.Consumer)} that
+     * obtains realm model from the session and puts it into session context before
+     * running the {@code what} task.
+     * @param <R>
+     * @param realmId
+     * @param what
+     * @return
+     */
     protected <R> R withRealm(String realmId, BiFunction<KeycloakSession, RealmModel, R> what) {
         return inComittedTransaction(session -> {
             final RealmModel realm = session.realms().getRealm(realmId);
@@ -327,4 +441,7 @@ public abstract class KeycloakModelTest {
         });
     }
 
+    protected boolean isUseSameKeycloakSessionFactoryForAllThreads() {
+        return false;
+    }
 }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java
@@ -42,9 +42,9 @@ import org.junit.Test;
 import static org.hamcrest.Matchers.hasItem;
 import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNull;
-import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assume.assumeThat;
 
@@ -79,6 +79,11 @@ public class UserModelTest extends KeycloakModelTest {
     @Override
     public void cleanEnvironment(KeycloakSession s) {
         s.realms().removeRealm(realmId);
+    }
+
+    @Override
+    protected boolean isUseSameKeycloakSessionFactoryForAllThreads() {
+        return true;
     }
 
     private Void addRemoveUser(KeycloakSession session, int i) {

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/AdminEventQueryTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/AdminEventQueryTest.java
@@ -22,14 +22,11 @@ import org.keycloak.events.EventStoreProvider;
 import org.keycloak.events.EventType;
 import org.keycloak.testsuite.model.KeycloakModelTest;
 import org.keycloak.testsuite.model.RequireProvider;
-import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.RealmModel;
 import java.util.stream.Collectors;
-import org.junit.After;
-import org.junit.Before;
 import org.junit.Test;
+import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
-import static org.junit.Assert.assertThat;
 
 /**
  *
@@ -38,39 +35,33 @@ import static org.junit.Assert.assertThat;
 @RequireProvider(EventStoreProvider.class)
 public class AdminEventQueryTest extends KeycloakModelTest {
 
-    private final KeycloakSession session = FACTORY.create();
-    private final EventStoreProvider eventStore = session.getProvider(EventStoreProvider.class);
 
     @Test
     public void testClear() {
-        eventStore.clear();
-    }
-
-    @Before
-    public void startTransaction() {
-        session.getTransactionManager().begin();
-    }
-
-    @After
-    public void stopTransaction() {
-        session.getTransactionManager().rollback();
+        inRolledBackTransaction(null, (session, t) -> {
+            EventStoreProvider eventStore = session.getProvider(EventStoreProvider.class);
+            eventStore.clear();
+        });
     }
 
     @Test
     public void testQuery() {
-        RealmModel realm = session.realms().createRealm("realm");
-        ClientConnection cc = new DummyClientConnection();
-        eventStore.onEvent(new EventBuilder(realm, null, cc).event(EventType.LOGIN).user("u1").getEvent());
-        eventStore.onEvent(new EventBuilder(realm, null, cc).event(EventType.LOGIN).user("u2").getEvent());
-        eventStore.onEvent(new EventBuilder(realm, null, cc).event(EventType.LOGIN).user("u3").getEvent());
-        eventStore.onEvent(new EventBuilder(realm, null, cc).event(EventType.LOGIN).user("u4").getEvent());
+        inRolledBackTransaction(null, (session, t) -> {
+            EventStoreProvider eventStore = session.getProvider(EventStoreProvider.class);
+            RealmModel realm = session.realms().createRealm("realm");
+            ClientConnection cc = new DummyClientConnection();
+            eventStore.onEvent(new EventBuilder(realm, null, cc).event(EventType.LOGIN).user("u1").getEvent());
+            eventStore.onEvent(new EventBuilder(realm, null, cc).event(EventType.LOGIN).user("u2").getEvent());
+            eventStore.onEvent(new EventBuilder(realm, null, cc).event(EventType.LOGIN).user("u3").getEvent());
+            eventStore.onEvent(new EventBuilder(realm, null, cc).event(EventType.LOGIN).user("u4").getEvent());
 
-        assertThat(eventStore.createQuery()
-          .firstResult(2)
-          .getResultStream()
-          .collect(Collectors.counting()),
-          is(2L)
-        );
+            assertThat(eventStore.createQuery()
+              .firstResult(2)
+              .getResultStream()
+              .collect(Collectors.counting()),
+              is(2L)
+            );
+        });
     }
 
     private static class DummyClientConnection implements ClientConnection {

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/Infinispan.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/Infinispan.java
@@ -36,12 +36,15 @@ import com.google.common.collect.ImmutableSet;
 import org.keycloak.timer.TimerProviderFactory;
 
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
 
 /**
  *
  * @author hmlnarik
  */
 public class Infinispan extends KeycloakModelParameters {
+
+    private static final AtomicInteger NODE_COUNTER = new AtomicInteger();
 
     static final Set<Class<? extends Spi>> ALLOWED_SPIS = ImmutableSet.<Class<? extends Spi>>builder()
       .add(CacheRealmProviderSpi.class)
@@ -67,7 +70,9 @@ public class Infinispan extends KeycloakModelParameters {
     public void updateConfig(Config cf) {
         cf.spi("connectionsInfinispan")
             .provider("default")
-              .config("embedded", "true");
+              .config("embedded", "true")
+              .config("clustered", "true")
+              .config("nodeName", "node-" + NODE_COUNTER.incrementAndGet());
     }
 
     public Infinispan() {

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java
@@ -1,0 +1,260 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.testsuite.model.session;
+
+import org.keycloak.models.AuthenticatedClientSessionModel;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.Constants;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.RealmProvider;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.UserProvider;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.models.UserSessionProvider;
+import org.keycloak.services.managers.RealmManager;
+import org.keycloak.testsuite.model.KeycloakModelTest;
+import org.keycloak.testsuite.model.RequireProvider;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ *
+ * @author hmlnarik
+ */
+@RequireProvider(UserProvider.class)
+@RequireProvider(RealmProvider.class)
+@RequireProvider(UserSessionProvider.class)
+public class OfflineSessionPersistenceTest extends KeycloakModelTest {
+
+    private static final int USER_COUNT = 50;
+    private static final int OFFLINE_SESSION_COUNT_PER_USER = 10;
+    
+    private String realmId;
+    private List<String> userIds;
+
+    @Override
+    public void createEnvironment(KeycloakSession s) {
+        RealmModel realm = prepareRealm(s, "realm");
+        this.realmId = realm.getId();
+
+        userIds = IntStream.range(0, USER_COUNT)
+          .mapToObj(i -> s.users().addUser(realm, "user-" + i))
+          .map(UserModel::getId)
+          .collect(Collectors.toList());
+    }
+
+    private static RealmModel prepareRealm(KeycloakSession s, String name) {
+        RealmModel realm = s.realms().createRealm(name);
+        realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
+        realm.setSsoSessionMaxLifespan(10 * 60 * 60);
+        realm.setSsoSessionIdleTimeout(1 * 60 * 60);
+        realm.setOfflineSessionMaxLifespan(365 * 24 * 60 * 60);
+        realm.setOfflineSessionIdleTimeout(30 * 24 * 60 * 60);
+        return realm;
+    }
+
+    @Override
+    public void cleanEnvironment(KeycloakSession s) {
+        new RealmManager(s).removeRealm(s.realms().getRealm(realmId));  // See https://issues.redhat.com/browse/KEYCLOAK-17876
+    }
+
+    @Test
+    public void testPersistenceSingleNodeDeleteRealm() {
+        String realmId2 = inComittedTransaction(session -> { return prepareRealm(session, "realm2").getId(); });
+        List<String> userIds2 = withRealm(realmId2, (session, realm) -> IntStream.range(0, USER_COUNT)
+          .mapToObj(i -> session.users().addUser(realm, "user2-" + i))
+          .map(UserModel::getId)
+          .collect(Collectors.toList())
+        );
+
+        try {
+            List<String> offlineSessionIds = createOfflineSessions(realmId, userIds);
+            assertOfflineSessionsExist(realmId, offlineSessionIds);
+
+            List<String> offlineSessionIds2 = createOfflineSessions(realmId2, userIds2);
+            assertOfflineSessionsExist(realmId2, offlineSessionIds2);
+
+            // Simulate server restart
+            reinitializeKeycloakSessionFactory();
+
+            withRealm(realmId2, (session, realm) -> new RealmManager(session).removeRealm(realm));
+
+            // Simulate server restart
+            reinitializeKeycloakSessionFactory();
+            assertOfflineSessionsExist(realmId, offlineSessionIds);
+        } finally {
+            withRealm(realmId2, (session, realm) -> realm == null ? false : new RealmManager(session).removeRealm(realm));
+        }
+    }
+
+    @Test
+    public void testPersistenceSingleNode() {
+        List<String> offlineSessionIds = createOfflineSessions(realmId, userIds);
+        assertOfflineSessionsExist(realmId, offlineSessionIds);
+
+        // Simulate server restart
+        reinitializeKeycloakSessionFactory();
+        assertOfflineSessionsExist(realmId, offlineSessionIds);
+    }
+
+    @Test(timeout = 90 * 1000)
+    public void testPersistenceMultipleNodesClientSessionAtSameNode() throws InterruptedException {
+        List<String> clientIds = withRealm(realmId, (session, realm) -> {
+            return IntStream.range(0, 5)
+              .mapToObj(cid -> (ClientModel) session.clients().addClient(realm, "client-" + cid))
+              .map(ClientModel::getId)
+              .collect(Collectors.toList());
+        });
+
+        // Shutdown factory -> enforce session persistence
+        closeKeycloakSessionFactory();
+
+        Map<String, List<String>> clientSessionIds = new ConcurrentHashMap<>();
+        inIndependentFactories(3, 30, () -> {
+            withRealm(realmId, (session, realm) -> {
+                // Create offline sessions
+                userIds.forEach(userId -> createOfflineSessions(session, realm, userId, offlineUserSession -> {
+                  List<String> innerClientSessionIds = IntStream.range(0, 5)
+                    .mapToObj(cid -> session.clients().getClientById(realm, clientIds.get(cid)))
+                    // TODO in the future: The following two lines are weird. Why an online client session needs to exist in order to create an offline one?
+                    .map(client -> session.sessions().createClientSession(realm, client, offlineUserSession))
+                    .map(clientSession -> session.sessions().createOfflineClientSession(clientSession, offlineUserSession))
+                    .map(AuthenticatedClientSessionModel::getId)
+                    .collect(Collectors.toList());
+                  clientSessionIds.put(offlineUserSession.getId(), innerClientSessionIds);
+                }));
+                return null;
+            });
+        });
+
+        reinitializeKeycloakSessionFactory();
+        inIndependentFactories(4, 30, () -> assertOfflineSessionsExist(realmId, clientSessionIds.keySet()));
+    }
+
+    @Test(timeout = 90 * 1000)
+    public void testPersistenceMultipleNodesClientSessionsAtRandomNode() throws InterruptedException {
+        List<String> clientIds = withRealm(realmId, (session, realm) -> {
+            return IntStream.range(0, 5)
+              .mapToObj(cid -> (ClientModel) session.clients().addClient(realm, "client-" + cid))
+              .map(ClientModel::getId)
+              .collect(Collectors.toList());
+        });
+        List<String> offlineSessionIds = createOfflineSessions(realmId, userIds);
+
+        // Shutdown factory -> enforce session persistence
+        closeKeycloakSessionFactory();
+
+        Map<String, List<String>> clientSessionIds = new ConcurrentHashMap<>();
+        AtomicInteger i = new AtomicInteger();
+        inIndependentFactories(3, 30, () -> {
+            for (int j = 0; j < USER_COUNT * 3; j ++) {
+                int index = i.incrementAndGet();
+                int oid = index % offlineSessionIds.size();
+                String offlineSessionId = offlineSessionIds.get(oid);
+                int cid = index % clientIds.size();
+                String clientSessionId = createOfflineClientSession(offlineSessionId, clientIds.get(cid));
+                clientSessionIds.computeIfAbsent(offlineSessionId, a -> new LinkedList<>()).add(clientSessionId);
+                if (index % 100 == 0) {
+                    reinitializeKeycloakSessionFactory();
+                }
+            }
+        });
+
+        reinitializeKeycloakSessionFactory();
+        assertOfflineSessionsExist(realmId, offlineSessionIds);
+    }
+
+    private String createOfflineClientSession(String offlineUserSessionId, String clientId) {
+        return withRealm(realmId, (session, realm) -> {
+            UserSessionModel offlineUserSession = session.sessions().getOfflineUserSession(realm, offlineUserSessionId);
+            ClientModel client = session.clients().getClientById(realm, clientId);
+            AuthenticatedClientSessionModel clientSession = session.sessions().createClientSession(realm, client, offlineUserSession);
+            return session.sessions().createOfflineClientSession(clientSession, offlineUserSession).getId();
+        });
+    }
+
+    @Test(timeout = 90 * 1000)
+    public void testPersistenceClientSessionsMultipleNodes() throws InterruptedException {
+        // Create offline sessions
+        List<String> offlineSessionIds = createOfflineSessions(realmId, userIds);
+
+        // Shutdown factory -> enforce session persistence
+        closeKeycloakSessionFactory();
+
+        inIndependentFactories(4, 30, () -> assertOfflineSessionsExist(realmId, offlineSessionIds));
+    }
+
+    /**
+     * Assert that all the offline sessions passed in the {@code offlineSessionIds} parameter exist
+     * @param factory
+     * @param offlineSessionIds
+     * @return
+     */
+    private Void assertOfflineSessionsExist(String realmId, Collection<String> offlineSessionIds) {
+        int foundOfflineSessions = withRealm(realmId, (session, realm) -> offlineSessionIds.stream()
+          .map(offlineSessionId -> session.sessions().getOfflineUserSession(realm, offlineSessionId))
+          .map(ous -> ous == null ? 0 : 1)
+          .reduce(0, Integer::sum));
+
+        assertThat(foundOfflineSessions, Matchers.is(USER_COUNT * OFFLINE_SESSION_COUNT_PER_USER));
+
+        return null;
+    }
+
+    // ***************** Helper methods *****************
+
+    /**
+     * Creates {@link #OFFLINE_SESSION_COUNT_PER_USER} offline sessions for every user from {@link #userIds}.
+     * @return Ids of the offline sessions
+     */
+    private List<String> createOfflineSessions(String realmId, List<String> userIds) {
+        return withRealm(realmId, (session, realm) ->
+          userIds.stream()
+            .flatMap(userId -> createOfflineSessions(session, realm, userId, us -> {}))
+            .map(UserSessionModel::getId)
+            .collect(Collectors.toList())
+        );
+    }
+
+    /**
+     * Creates {@link #OFFLINE_SESSION_COUNT_PER_USER} offline sessions for {@code userId} user.
+     */
+    private Stream<UserSessionModel> createOfflineSessions(KeycloakSession session, RealmModel realm, String userId, Consumer<? super UserSessionModel> alterUserSession) {
+        return IntStream.range(0, OFFLINE_SESSION_COUNT_PER_USER)
+          .mapToObj(sess -> createOfflineSession(session, realm, userId, sess))
+          .peek(alterUserSession == null ? us -> {} : us -> alterUserSession.accept(us));
+    }
+
+    private UserSessionModel createOfflineSession(KeycloakSession session, RealmModel realm, String userId, int sessionIndex) {
+        final UserModel user = session.users().getUserById(realm, userId);
+        UserSessionModel us = session.sessions().createUserSession(realm, user, "un" + sessionIndex, "ip1", "auth", false, null, null);
+        return session.sessions().createOfflineUserSession(us);
+    }
+
+}

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.keycloak.testsuite.model;
+package org.keycloak.testsuite.model.session;
 
 import org.junit.Assert;
 import org.junit.Test;
@@ -39,6 +39,8 @@ import org.keycloak.services.managers.UserSessionManager;
 import java.util.List;
 import java.util.stream.Collectors;
 
+import org.keycloak.testsuite.model.KeycloakModelTest;
+import org.keycloak.testsuite.model.RequireProvider;
 import static org.hamcrest.core.Is.is;
 import static org.junit.Assert.assertThat;
 

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.keycloak.testsuite.model;
+package org.keycloak.testsuite.model.session;
 
 import org.junit.Assert;
 import org.junit.Test;
@@ -53,6 +53,8 @@ import static org.hamcrest.MatcherAssert.assertThat;
 import org.keycloak.models.Constants;
 import org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory;
 import org.hamcrest.Matchers;
+import org.keycloak.testsuite.model.KeycloakModelTest;
+import org.keycloak.testsuite.model.RequireProvider;
 
 /**
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionProviderModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionProviderModelTest.java
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.keycloak.testsuite.model;
+package org.keycloak.testsuite.model.session;
 
 import org.junit.Assert;
 import org.junit.Test;
@@ -41,8 +41,10 @@ import java.util.List;
 import java.util.UUID;
 import java.util.concurrent.atomic.AtomicReference;
 
-import static org.keycloak.testsuite.model.UserSessionPersisterProviderTest.createClients;
-import static org.keycloak.testsuite.model.UserSessionPersisterProviderTest.createSessions;
+import org.keycloak.testsuite.model.KeycloakModelTest;
+import org.keycloak.testsuite.model.RequireProvider;
+import static org.keycloak.testsuite.model.session.UserSessionPersisterProviderTest.createClients;
+import static org.keycloak.testsuite.model.session.UserSessionPersisterProviderTest.createSessions;
 
 /**
  * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionProviderOfflineModelTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionProviderOfflineModelTest.java
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.keycloak.testsuite.model;
+package org.keycloak.testsuite.model.session;
 
 import org.junit.Assert;
 import org.junit.Test;
@@ -44,6 +44,8 @@ import java.util.Map;
 import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
+import org.keycloak.testsuite.model.KeycloakModelTest;
+import org.keycloak.testsuite.model.RequireProvider;
 
 /**
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>

--- a/testsuite/model/src/test/resources/log4j.properties
+++ b/testsuite/model/src/test/resources/log4j.properties
@@ -46,3 +46,9 @@ log4j.logger.org.keycloak.connections.jpa.DefaultJpaConnectionProviderFactory=de
 #log4j.logger.org.keycloak.models.map.transaction=debug
 #
 #log4j.logger.org.keycloak.STACK_TRACE=trace
+
+#log4j.logger.org.keycloak.models.sessions.infinispan=trace
+#
+#log4j.logger.org.hibernate.type=trace
+#log4j.logger.org.hibernate.SQL=debug 
+

